tests/extmod/machine_spi_rate.py

```diff
@@ -0,0 +1,152 @@
+# Test machine.SPI data transfer rates
+import sys
+
+
+try:
+    import time
+    from machine import Pin, SPI
+except ImportError:
+    print("SKIP")
+    raise SystemExit
+
+MAX_DELTA_MS = 8
+
+# Configure pins based on the port/board details.
+# Values are tuples of (spi_id, sck, mosi, miso)
+if "pyboard" in sys.platform:
+    spi_instances = (
+        (1, None, None, None),  # "explicit choice of sck/mosi/miso is not implemented"
+        (2, None, None, None),
+    )
+elif "rp2" in sys.platform:
+    spi_instances = ((0, Pin(18), Pin(19), Pin(16)),)
+elif "esp32" in sys.platform:
+    spi_instances = [
+        (1, Pin(18), Pin(19), Pin(21)),
+    ]
+    if "ESP32C3" not in str(sys.implementation):
+        spi_instances.append((2, Pin(18), Pin(19), Pin(21)))
+else:
+    print("Please add support for this test on this platform.")
+    raise SystemExit
+
+
+def get_real_baudrate(spi):
+    # Return the 'real' baudrate for a SPI object, from parsing 'print' output
+    # i.e.
+    # SPI(id=1, baudrate=500000, polarity=0, phase=0, bits=8, firstbit=0, sck=14, mosi=13, miso=12)
+    #
+    # Note the 'real' rate may be quite different to the requested rate, i.e.
+    # on ports where the SPI hardware only supports power of 2 clock dividers.
+    #
+    # Implementation is somewhat fiddly and inefficient to avoid dependency on
+    # 're' module,
+    s = str(spi)
```

Comment:
@dpgeorge I went with str() here rather than the suggestion of printing to StringIO because it seems like this should work everywhere (string constructor calls through to mp_obj_print_helper()`.)

Reply:
Oh yeah, that's much simpler!